### PR TITLE
Add ThreadLocal utility class

### PR DESCRIPTION
**Pull request by vsbogd**
Add ThreadLocal utility class to allow threads using local instances of classes without contention.
https://github.com/vsbogd/atomspace/tree/parallel-pattern-matcher depends on it.

ThreadLocal usage example: https://github.com/vsbogd/atomspace/blob/cda8784ae7fdb754f2460157ae2cdf0049738212/opencog/query/DefaultPatternMatchCB.h#L156
Logger usage example: https://github.com/vsbogd/atomspace/blob/cda8784ae7fdb754f2460157ae2cdf0049738212/opencog/query/InitiateSearchCB.cc#L396

### CODE REVIEW

File: opencog/util/CMakeLists.txt [hunk 1]
```diff
@@ -119,6 +120,7 @@ INSTALL(FILES
 	StringTokenizer.h
 	tbb.h
 	tree.h
+	ThreadLocal.h
```
Comment (ngeiswei) [inline opencog/util/CMakeLists.txt]: Total minor detail but there are some trailing spaces.

Comment (vsbogd) [reply opencog/util/CMakeLists.txt]: Fixed, thanks

**Comment by ngeiswei** [comment]
@vsbogd You should update `Logger::set` so that `threadIdEnabled` gets properly assigned, I think.

File: opencog/util/Logger.cc [hunk 2]
```diff
@@ -273,7 +274,8 @@ Logger::Logger(const std::string &fname, Logger::Level level, bool tsEnabled)
 }
 
 Logger::Logger(const Logger& log)
-    : error(*this), warn(*this), info(*this), debug(*this), fine(*this)
+    : error(*this), warn(*this), info(*this), debug(*this), fine(*this),
+      threadIdEnabled(false)
```
Comment (ngeiswei) [inline opencog/util/Logger.cc]: Shouldn't `threadIdEnabled` just be assigned the value of `log.threadIdEnabled` instead of being initialized to false?

**Comment by linas** [comment]
Can you explain this? Why is the standard thread-local storage, as supplied by standard c++, not sufficient to do everything you want?

**Comment by linas** [comment]
Also, the logger is already thread-safe; so I am confused about why you feel you need a thread-local variant.

**Comment by linas** [comment]
OK, now some comments re thread-local.

So, one standard idiom for storage that gets automatically cleaned up is to put that storage on the stack.  When the stack unwinds, the storage is cleaned up automatically.  I don't see why this is not sufficient in this case.

Separately, I kind of would have liked to talk to you about parallelizing the pattern matcher before you started writing code, instead of after. The "natural", simplest, easiest place to parallelize would be to make the initiate-search loops parallel, and then allow each search to run in a distinct thread.  This, it would seem to me, to be a natural place to set up the per-thread stack: its located "at the root" of the search, and so is guaranteed to be alive, until that thread finishes. 

Doing this would seem to obviate the need for some per-thread storage that shows up "from thin air", in the middle of some random code doing some random thing.  I'm rather nervous about the debuggability and correctness of what you are proposing to do. 

Anyway, this pull req is the wrong place to have this discussion.

**Comment by vsbogd** [comment]
Extracted logging functionality to https://github.com/opencog/cogutil/pull/110, closing this PR

**Pull request closed without merging**